rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,837 @@
+runtime.
+
+## Size Reduction
+
+Since we want to run on a mobile platform, we need to look for any opportunity
+to cut down size. First of all, we remove dependency on current TensorFlow
+runtime (for e.g. we no longer depend on `NodeDef` and `OpDef` protos). We are
+also looking at ways to reduce large size contributions of
+[absl libraries](https://github.com/abseil/abseil-cpp/tree/master/absl) and
+[protos](https://github.com/protocolbuffers/protobuf).
+
+### Protos
+
+We are currently investigating the following options:
+
+*   Switch to [micropb](https://github.com/protocolbuffers/upb). This proto
+    implementation provides C interfaces and is more compact.
+*   Remove dependency on protos.
+
+### ABSL
+
+We can hide ABSL references behind aliases (see
+[tensorflow::StringPiece](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/stringpiece.h;l=33;drc=af7fd02ca40f362c4ac96dd064d6a2224b65d784)
+for example) to make it easier to replace all references to save binary size.
+
+@gunhan is also starting an effort to define a library of STL utilities that
+helps us cut down on binary size.
+
+## Selecting which kernels to register
+
+We want to add a script to build configurations that can determine required
+kernels based on a model. We would then only build these kernels. For now, we will only support selective registration when building from source.
+
+Script details still need to be worked out.
+
+### Alternatives Considered
+
+The main alterantive to TFRT Kernel Fallback is TFRT Runtime Fallback. TFRT
+Runtime Fallback will call TensorFlow Eager C API (corresponding RFC should be
+published soon). Main trade offs between the two fallbacks are described in the
+table below:
+
+Property    | TFRT Kernel Fallback                           | TFRT Runtime Fallback
+----------- | ---------------------------------------------- | ---------------------
+Generality  | Support subset of ops (for e.g. no resources*) | Support all ops
+Scalability | Requires per-kernel updates                    | No kernel changes
+Performance | Lower overhead                                 | Higher overhead
+Binary size | Lower (no TF runtime)                          | Higher
+
+\* Long term we might support resources, but we consider them lower priority
+due to significant work involved.
+
+### Performance Implications
+
+*   Slow down due to adding inheritance for `OpKernelContext` and
+    `OpKernelConsturction`.
+*   Speed up for lighter weight kernel calls.
+
+We will run benchmarks to check performance numbers as we work on the
+implementation.
+
+### Dependencies
+
+No new dependencies.
+
+### Engineering Impact
+
+*   Build / startup time / binary size will be impacted by additional code added
+    to implement Kernel Fallback. At the same time one of the goals of Kernel
+    Fallback is to provide a lower-binary-size way to run existing TensorFlow
+    kernels in TF Lite.
```

Comment:
This is related to my point above. The mobile related goals span also TFRT:
Short term we want to run kernels on TFLite Select, but also run them on mobile via TFRT.

Reply:
I removed reference to TF Lite and instead changed it to:
"At the same time one of the goals of Kernel Fallback is to provide a lower-binary-size way to run existing TensorFlow kernels on mobile."
Specifically, this document is only looking at TFRT -> TF kernel interaction on mobile. Whether this path will be triggered purely from TFRT or if a similar approach would be triggered from TF Lite Select is outside of the scope.